src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,1041 @@
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers);
+  _sampler->handle_timer_signal(info, context);
+  Atomic::dec(&_sampler->_active_signal_handlers);
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(siginfo_t* info, void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrThreadLocal* jtl = jt->jfr_thread_local();
+  if (jtl == nullptr) {
+    return;
+  }
+  if (jt->thread_state() != _thread_in_native) {
+    jtl->set_wants_cpu_time_out_of_safepoint_recording(false);
+  }
+  if (!jtl->acquire_cpu_time_jfr_enqueue_lock()) {
+    jtl->increment_cpu_time_lost_samples();
+    return;
+  }
+  NoResourceMark nrm;
+  JfrCPUTimeTrace* trace = _frame_store.get_fresh_trace();
+  if (trace != nullptr) {
+    // the sampling period might be too low for the current Linux configuration
+    // so samples might be skipped and we have to compute the actual period
+    int64_t period = get_sampling_period() * (info->si_overrun + 1);
+    trace->record_trace(jt, context, period);
+    if (jtl->cpu_time_jfr_stack().enqueue(trace)) {
+      jtl->set_has_cpu_time_jfr_events(true);
+      SafepointMechanism::arm_local_poll_release(jt);
+    } else {
+      jtl->increment_cpu_time_lost_samples();
+    }
+  } else {
+    jtl->increment_cpu_time_lost_samples();
+  }
+
+  if (jt->thread_state() == _thread_in_native) {
+    u4 own_size = jtl->cpu_time_jfr_stack().size();
+    u4 own_capacity = jtl->cpu_time_jfr_stack().capacity();
+    u4 fresh_size = this->_frame_store.fresh_frames();
+    u4 fresh_capacity = this->_frame_store.capacity();
+    if (own_size > own_capacity / 2 || (own_size > own_capacity / 10 && fresh_size > fresh_capacity / 2)) {
+      // we're running out of space or out of fresh frames
+      jtl->set_wants_cpu_time_out_of_safepoint_recording(true);
+      trigger_out_of_safepoint_recording();
+    }
+  }
+
+  jtl->release_cpu_time_jfr_queue_lock();
+}
+
+static const int SIG = SIGPROF;
+
+static void set_timer_time(timer_t timerid, int64_t period_nanos) {
+  struct itimerspec its;
+  if (period_nanos == 0) {
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+  } else {
+    its.it_interval.tv_sec = period_nanos / NANOSECS_PER_SEC;
+    its.it_interval.tv_nsec = period_nanos % NANOSECS_PER_SEC;
+  }
+  its.it_value = its.it_interval;
+  if (timer_settime(timerid, 0, &its, nullptr) == -1) {
+    warning("Failed to set timer for thread sampling: %s", os::strerror(os::get_last_error()));
+  }
+}
+
+bool JfrCPUTimeThreadSampler::create_timer_for_thread(JavaThread* thread, timer_t& timerid) {
+  if (thread->osthread() == nullptr || thread->osthread()->thread_id() == 0){
+    return false;
+  }
+  timer_t t;
+  struct sigevent sev;
+  sev.sigev_notify = SIGEV_THREAD_ID;
+  sev.sigev_signo = SIG;
+  sev.sigev_value.sival_ptr = &t;
+  ((int*)&sev.sigev_notify)[1] = thread->osthread()->thread_id();
```

Comment:
Is there a reason why not use `sev._sigev_un._tid` directly if this is Linux only?

Reply:
Good catch.

Reply:
Are you sure the field is exposed in headers on all Linux distributions?
IIRC, it was not on Alpine.